src/Prefetch.cpp

```diff
@@ -220,8 +206,122 @@ class InjectPrefetch : public IRMutator {
+            internal_assert(base);
+
+            vector<string> index_names;
+            Expr offset = 0;
+            for (size_t i = max_arg_size; i < call->args.size(); i += 2) {
+                Expr stride = call->args[i+1];
+                string index_name = "prefetch_reduce_" + base->name + "." + std::to_string((i-1)/2);
+                index_names.push_back(index_name);
+                offset += Variable::make(Int(32), index_name) * stride;
+            }
+            Expr new_base = Load::make(base->type, base->name, simplify(base->index + offset),
+                                       base->image, base->param, base->predicate);
+            Expr new_base_addr = Call::make(Handle(), Call::address_of, {new_base}, Call::Intrinsic);
+
+            vector<Expr> args = {new_base_addr};
+            for (size_t i = 1; i < max_arg_size; ++i) {
+                args.push_back(call->args[i]);
+            }
+
+            stmt = Evaluate::make(Call::make(call->type, Call::prefetch, args, Call::Intrinsic));
+            for (size_t i = 0; i < index_names.size(); ++i) {
+                stmt = For::make(index_names[i], 0, call->args[(i+max_dim)*2 + 1],
+                                 ForType::Serial, DeviceAPI::None, stmt);
+            }
+            debug(5) << "\nReduce prefetch to " << max_dim << " dim:\n"
+                     << "Before:\n" << Expr(call) << "\nAfter:\n" << stmt << "\n";
+        }
     }
 
+public:
+    ReducePrefetchDimension(size_t dim) : max_dim(dim) {}
+};
+
+class SplitPrefetch : public IRMutator {
+    using IRMutator::visit;
+
+    Expr max_byte_size;
+
+    void visit(const Evaluate *op) {
+        IRMutator::visit(op);
+        op = stmt.as<Evaluate>();
+        internal_assert(op);
+        const Call *call = op->value.as<Call>();
+
+        // If the prefetched data is larger than 'max_byte_size', we need to
+        // tile the prefetch.
+
+        if (call && call->is_intrinsic(Call::prefetch)) {
+            const Call *base_addr = call->args[0].as<Call>();
+            internal_assert(base_addr && base_addr->is_intrinsic(Call::address_of));
+            const Load *base = base_addr->args[0].as<Load>();
+            internal_assert(base);
+
+            int elem_size = base->type.bytes();
+
+            vector<string> index_names;
+            vector<Expr> extents;
+            Expr offset = 0;
+            for (size_t i = 1; i < call->args.size(); i += 2) {
+                Expr extent = call->args[i];
+                Expr stride = call->args[i+1];
+                Expr stride_bytes = stride * elem_size;
+
+                string index_name = "prefetch_split_" + base->name + "." + std::to_string((i-1)/2);
+                index_names.push_back(index_name);
+                offset += Variable::make(Int(32), index_name) * stride_bytes;
+
+                Expr outer_extent = simplify((extent * stride_bytes + max_byte_size - 1)/max_byte_size);
```

Comment:
If there is a 2D prefetch with a large stride (but small extent), isn't this going to generate tons of unnecessary prefetches?

This logic makes sense for 1D prefetches, but I think it might not be what we want for multi-dimensional prefetches.